lean/deps/typed_smt2/src/galois/smt2/basic.lean

```diff
@@ -0,0 +1,185 @@
+def inhabited_type.uninterpreted : inhabited_type := { type := ℕ, value := 0 }
+
+/--
+This defines an interpretation of sorts.
+
+It maps each sort to an inhabited type, and the mapping for bool must
+be Prop.
+
+As there are an unbounded number of sorts, implementations should map
+the ones they recognize to the appropriate type, then map the others
+to `inhabited_type.uninterpreted` to represent uninterpreted types.
+
+We use a function rather than finite map to deal with polymorphic
+types such as bitvectors.
+-/
+structure logic :=
+(eval : sort → inhabited_type)
+(bool_eqn : (eval bool).type = Prop)
+
+namespace logic
+
+/-- Return the type associated with the given sort or Prop if not defined -/
+def type_of (l:logic) (s:sort) : Type := (l.eval s).type
+
+/-- Assert proposition is assigned to Booleans -/
+def prop_is_bool_type (l:logic) : l.type_of bool = Prop := l.bool_eqn
+
+/-- Return the type associated with the given sort or Prop if not defined -/
+def value_of (l:logic) (s:sort) : l.type_of s := (l.eval s).value
+
+/-- @fun_type args res@ dReturn the type assigned to functions. -/
+def fun_domain (l:logic) : list sort → sort → Type
+| [] res := l.type_of res
+| (h::r) tp := l.type_of h → fun_domain r tp
+
+/-- Default value for a function -/
+def fun_default_value (l:logic) : Π(args:list sort) (res:sort), l.fun_domain args res
+| [] res := l.value_of res
+| (h::r) tp := λ(x : l.type_of h), fun_default_value r tp
+
+/-- Coercision from prop to l.type_of bool -/
+instance bool_coe (l:logic) : has_coe Prop (l.type_of bool) :=
+{ coe := λ(x:Prop), eq.rec x l.prop_is_bool_type.symm }
+
+end logic
+
+------------------------------------------------------------------------
+-- Model
+
+/--
+Provides a valuation of symbols in a SMT formula.  This is parameterized by a logic, which
+defines how SMTLIB sorts should be mapped to Lean types.
+
+This maps symbols (which could be functions) to their type information and a valuation with
+the write domain according to the logic.
+-/
+def model (l:logic) : Type := rbmap symbol (Σ(args:list sort) (s:sort), l.fun_domain args s)
+
+instance (l:logic) : has_emptyc (model l) :=
+{ emptyc := mk_rbmap _ _ }
+
+/--
+A term in SMTLIB representation.
+
+Our term represnetation is extensible so that one can easily construct
+terms with arbitrary sexpression respresentations and interpretations.
+
+This is designed to make it easy to introduce new theories without
+changing the core of how the language works, but has the risk that one
+can give an incorrect interpretation to an expression.
+
+We recommend users only construct terms manually if they must, and
+otherwise use existing definitions.
+-/
+
+structure term (s : sort) :=
+(repr : sexpr)
+(interp : Π{l:logic}, model l → l.type_of s)
+
+namespace model
+
+section
+parameter {l:logic}
+
+/-- Add a binding -/
+def bind (ctx:model l)  (nm:symbol) {args: list sort} {r:sort} (v:l.fun_domain args r) : model l :=
```

Comment:
What is the conceptual difference between `bind` and `bind_const`?

Reply:
Clarified comments.